pkg/wireguard/wireguard.go

```diff
@@ -494,3 +524,37 @@ func asUDP(addr netip.AddrPort) *net.UDPAddr {
 		Zone: addr.Addr().Zone(),
 	}
 }
+
+func runPeersDrainer(ctx context.Context, peers PeerSource) (chan []PeerEvent, func([]PeerEvent)) {
+	pool := sync.Pool{
+		New: func() any { return make([]PeerEvent, 0, 100) },
+	}
+
+	resultCh := make(chan []PeerEvent)
+
+	go func() {
+		slc := pool.Get().([]PeerEvent) //nolint:errcheck,forcetypeassert
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case peerEvent := <-peers.EventCh():
+				slc = append(slc, peerEvent)
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case resultCh <- slc:
+				slc = pool.Get().([]PeerEvent) //nolint:errcheck,forcetypeassert
+			default:
```

Comment:
I'm confused here a bit, if `resultCh` is not ready to receive, we go to the `for{}` loop once again, and we don't try to send anymore.

So if there is a single peer event coming, and `resultCh` wasn't ready, we will fall through to line 542, and never try to send ever? What if `peerEvent` never comes?

Reply:
Good catch, I didn't thought about that. We can actually simplify the code but using `sending to nil channel always blocks` idiom. [Relevant changes](https://github.com/siderolabs/siderolink/compare/e5d290e5badd11e4257b894eaea703a205ed45f7..a936b60645267d2e7320083b402df5ad19de76f5) and try to send the events always.